glusterd2/commands/devices/adddevice.go

```diff
@@ -0,0 +1,112 @@
+package devicecommands
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/pborman/uuid"
+	"net/http"
+
+	"github.com/coreos/etcd/clientv3"
+	device "github.com/gluster/glusterd2/glusterd2/device"
+	restutils "github.com/gluster/glusterd2/glusterd2/servers/rest/utils"
+	"github.com/gluster/glusterd2/glusterd2/store"
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/pkg/api"
+	log "github.com/sirupsen/logrus"
+)
+
+func deviceAddHandler(w http.ResponseWriter, r *http.Request) {
+	// Collect inputs from URLi
+	ctx := r.Context()
+	req := new(device.AddDeviceReq)
+	if err := restutils.UnmarshalRequest(r, req); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "", api.ErrCodeDefault)
+		return
+	}
+	nodeIDRaw := req.NodeID
+	deviceName := req.DeviceName
+	var deviceinfo device.Info
+	deviceinfo.DeviceName = deviceName
+	nodeID := nodeIDRaw
+	if nodeID == nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Invalid Node ID", api.ErrCodeDefault)
+		return
+	}
+	deviceinfo.NodeID = nodeID
+	_, err := store.Store.Get(context.TODO(), deviceinfo.NodeID.String())
+	if err != nil {
+
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "Node Id not found in store", api.ErrCodeDefault)
+		return
+	}
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
+
+	lock, unlock, err := transaction.CreateLockSteps(deviceinfo.NodeID.String())
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "Unable to acquire lock", api.ErrCodeDefault)
+		return
+	}
+
+	nodes := make([]uuid.UUID, 0)
+	nodes = append(nodes, deviceinfo.NodeID)
+
+	txn.Nodes = nodes
+	txn.Steps = []*transaction.Step{
+		lock,
+		{
+			DoFunc: "prepare-device.Commit",
+			Nodes:  txn.Nodes,
+		},
+		unlock,
+	}
+	txn.Ctx.Set("nodeid", deviceinfo.NodeID.String())
+	txn.Ctx.Set("devicename", deviceinfo.DeviceName)
+
+	err = txn.Do()
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "Transaction Failed", api.ErrCodeDefault)
+		return
+	}
+	// update device state
+	deviceinfo.State = device.DeviceEnabled
+	deviceJSON, err := json.Marshal(deviceinfo)
+	if err != nil {
+		log.WithField("error", err).Error("Failed to marshal the DeviceInfo object")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "Failed to marshal the DeviceInfo object", api.ErrCodeDefault)
+		return
+	}
+
+	deviceDetails, _ := store.Store.Get(context.TODO(), "devices/"+nodeID.String(), clientv3.WithPrefix())
+
```

Comment:
What exactly is being done here ?

Can you describe what this PR does and the API does on the PR or commit message ?